File: vignettes/11_resilience/python/11_resilience.py
"""
Resilience — Retry and rate limiting (Python)

This sample demonstrates production-grade resilience patterns: retry with
exponential backoff and token-bucket rate limiting. It mirrors the Julia
vignette 11_resilience.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
import random
import time
from collections.abc import Callable
from typing import Any

from agent_framework.ollama import OllamaChatClient


# ── Retry Configuration ──────────────────────────────────────────────────────


class RetryConfig:
    """Configuration for retry with exponential backoff."""

    def __init__(
        self,
        max_retries: int = 3,
        initial_delay: float = 1.0,
        max_delay: float = 60.0,
        multiplier: float = 2.0,
        jitter: float = 0.1,
        retryable_status_codes: list[int] | None = None,
    ):
        self.max_retries = max_retries
        self.initial_delay = initial_delay
        self.max_delay = max_delay
        self.multiplier = multiplier
        self.jitter = jitter
        self.retryable_status_codes = retryable_status_codes or [429, 500, 502, 503, 504]

    def compute_delay(self, attempt: int) -> float:
        """Compute backoff delay with jitter for a given attempt."""
        base_delay = min(self.initial_delay * (self.multiplier ** attempt), self.max_delay)
        jitter_amount = base_delay * self.jitter * (2 * random.random() - 1)
        return max(0.0, base_delay + jitter_amount)

    def is_retryable(self, error: Exception) -> bool:
        """Check if an error should trigger a retry."""
        error_msg = str(error).lower()
        retryable_keywords = ["timeout", "rate limit", "connection", "temporary"]
        for code in self.retryable_status_codes:
            if str(code) in error_msg:
                return True
        return any(kw in error_msg for kw in retryable_keywords)


# ── Token Bucket Rate Limiter ────────────────────────────────────────────────


class TokenBucketRateLimiter:
    """Thread-safe token bucket rate limiter."""

    def __init__(self, requests_per_second: float = 10.0, burst: int = 10):
        self.capacity = float(burst)
        self.tokens = float(burst)
        self.refill_rate = requests_per_second
        self.last_refill = time.monotonic()

    def _refill(self) -> None:
        now = time.monotonic()
        elapsed = now - self.last_refill
        self.tokens = min(self.capacity, self.tokens + elapsed * self.refill_rate)
        self.last_refill = now

    async def acquire(self, timeout: float = 30.0) -> bool:
        """Wait until a token is available or timeout."""
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            self._refill()
            if self.tokens >= 1.0:
                self.tokens -= 1.0
                return True
            wait_time = min(1.0 / self.refill_rate, deadline - time.monotonic())
            if wait_time > 0:
                await asyncio.sleep(wait_time)
        return False


# ── Resilient Agent Wrapper ──────────────────────────────────────────────────


async def run_with_retry(
    agent_func: Callable[..., Any],
    *args: Any,
    config: RetryConfig | None = None,
    **kwargs: Any,
) -> Any:
    """Run an async agent function with retry logic."""
    config = config or RetryConfig()
    last_error: Exception | None = None

    for attempt in range(config.max_retries + 1):
        try:
            return await agent_func(*args, **kwargs)
        except Exception as e:
            last_error = e
            if attempt == config.max_retries or not config.is_retryable(e):
                raise
            delay = config.compute_delay(attempt)
            print(f"  Retry {attempt + 1}/{config.max_retries}: "
                  f"waiting {delay:.2f}s after {type(e).__name__}")
            await asyncio.sleep(delay)

    raise last_error  # type: ignore[misc]


async def main() -> None:
    client = OllamaChatClient(host="http://localhost:11434", model_id="qwen3:8b")
    agent = client.as_agent(
        name="ResilientAgent",
        instructions="You are a helpful assistant. Keep answers brief.",
    )

    # ── 1. Retry Configuration ───────────────────────────────────────────
    print("=== Retry Configuration ===")
    config = RetryConfig(
        max_retries=3,
        initial_delay=1.0,
        multiplier=2.0,
        jitter=0.1,
    )
    for attempt in range(4):
        delay = config.compute_delay(attempt)
        print(f"  Attempt {attempt}: delay = {delay:.2f}s")

    # ── 2. Rate Limiter ──────────────────────────────────────────────────
    print("\n=== Rate Limiter ===")
    limiter = TokenBucketRateLimiter(requests_per_second=5.0, burst=10)
    acquired = await limiter.acquire(timeout=5.0)
    print(f"  Token acquired: {acquired}")

    # ── 3. Resilient Agent Call ──────────────────────────────────────────
    print("\n=== Resilient Agent Call ===")
    result = await run_with_retry(
        agent.run,
        "What is the speed of light?",
        config=config,
    )
    print(f"  Agent: {result}")


if __name__ == "__main__":
    asyncio.run(main())

File: vignettes/11_resilience/csharp/11_resilience.cs
// Resilience — Retry and rate limiting (C#)
//
// This sample demonstrates production-grade resilience patterns: retry with
// exponential backoff and token-bucket rate limiting. It mirrors the Julia
// vignette 11_resilience.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ── 1. Retry Configuration ──────────────────────────────────────────────────

Console.WriteLine("=== Retry Configuration ===");

var retryConfig = new RetryConfig(
    MaxRetries: 3,
    InitialDelay: TimeSpan.FromSeconds(1),
    MaxDelay: TimeSpan.FromSeconds(60),
    Multiplier: 2.0,
    Jitter: 0.1
);

for (int attempt = 0; attempt < 4; attempt++)
{
    var delay = retryConfig.ComputeDelay(attempt);
    Console.WriteLine($"  Attempt {attempt}: delay = {delay.TotalSeconds:F2}s");
}

// ── 2. Rate Limiter ─────────────────────────────────────────────────────────

Console.WriteLine("\n=== Rate Limiter ===");

var limiter = new TokenBucketRateLimiter(requestsPerSecond: 5.0, burst: 10);
bool acquired = await limiter.AcquireAsync(timeout: TimeSpan.FromSeconds(5));
Console.WriteLine($"  Token acquired: {acquired}");

// ── 3. Resilient Agent Call ─────────────────────────────────────────────────

Console.WriteLine("\n=== Resilient Agent Call ===");

AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        name: "ResilientAgent",
        instructions: "You are a helpful assistant. Keep answers brief.");

var response = await RetryHelper.RunWithRetryAsync(
    async () => await agent.RunAsync("What is the speed of light?"),
    retryConfig
);
Console.WriteLine($"  Agent: {response.Text}");

// ── Supporting Types ────────────────────────────────────────────────────────

/// <summary>Configuration for retry with exponential backoff.</summary>
record RetryConfig(
    int MaxRetries = 3,
    TimeSpan? InitialDelay = null,
    TimeSpan? MaxDelay = null,
    double Multiplier = 2.0,
    double Jitter = 0.1)
{
    private static readonly Random s_random = new();

    public TimeSpan ComputeDelay(int attempt)
    {
        var initial = InitialDelay ?? TimeSpan.FromSeconds(1);
        var max = MaxDelay ?? TimeSpan.FromSeconds(60);
        double baseDelay = Math.Min(
            initial.TotalSeconds * Math.Pow(Multiplier, attempt),
            max.TotalSeconds);
        double jitterAmount = baseDelay * Jitter * (2 * s_random.NextDouble() - 1);
        return TimeSpan.FromSeconds(Math.Max(0, baseDelay + jitterAmount));
    }

    public bool IsRetryable(Exception ex)
    {
        string msg = ex.Message.ToLowerInvariant();
        int[] retryableCodes = [429, 500, 502, 503, 504];
        string[] retryableKeywords = ["timeout", "rate limit", "connection", "temporary"];
        return retryableCodes.Any(c => msg.Contains(c.ToString()))
            || retryableKeywords.Any(kw => msg.Contains(kw));
    }
}

/// <summary>Token bucket rate limiter for controlling request throughput.</summary>
class TokenBucketRateLimiter
{
    private readonly double _capacity;
    private double _tokens;
    private readonly double _refillRate;
    private DateTime _lastRefill;

    public TokenBucketRateLimiter(double requestsPerSecond = 10.0, int burst = 10)
    {
        _capacity = burst;
        _tokens = burst;
        _refillRate = requestsPerSecond;
        _lastRefill = DateTime.UtcNow;
    }

    public async Task<bool> AcquireAsync(TimeSpan? timeout = null)
    {
        var deadline = DateTime.UtcNow + (timeout ?? TimeSpan.FromSeconds(30));
        while (DateTime.UtcNow < deadline)
        {
            Refill();
            if (_tokens >= 1.0)
            {
                _tokens -= 1.0;
                return true;
            }
            var waitTime = Math.Min(1.0 / _refillRate, (deadline - DateTime.UtcNow).TotalSeconds);
            if (waitTime > 0) await Task.Delay(TimeSpan.FromSeconds(waitTime));
        }
        return false;
    }

    private void Refill()
    {
        var now = DateTime.UtcNow;
        var elapsed = (now - _lastRefill).TotalSeconds;
        _tokens = Math.Min(_capacity, _tokens + elapsed * _refillRate);
        _lastRefill = now;
    }
}

/// <summary>Helper for running async functions with retry logic.</summary>
static class RetryHelper
{
    public static async Task<T> RunWithRetryAsync<T>(Func<Task<T>> func, RetryConfig config)
    {
        Exception? lastError = null;
        for (int attempt = 0; attempt <= config.MaxRetries; attempt++)
        {
            try
            {
                return await func();
            }
            catch (Exception ex)
            {
                lastError = ex;
                if (attempt == config.MaxRetries || !config.IsRetryable(ex))
                    throw;
                var delay = config.ComputeDelay(attempt);
                Console.WriteLine($"  Retry {attempt + 1}/{config.MaxRetries}: "
                    + $"waiting {delay.TotalSeconds:F2}s after {ex.GetType().Name}");
                await Task.Delay(delay);
            }
        }
        throw lastError!;
    }
}
